<a href="https://colab.research.google.com/github/DulainNimthakaJayasumana/Learning-ML/blob/main/EvaluateModelWeek2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Regression Predictive Model

In [2]:
import pandas as pd

# Load the dataset
data = pd.read_csv('/content/slr.csv')

# View the first few rows
data.head()

chirps per second  temperature (F)
0          20.000000        88.599998
1          16.000000        71.599998
2          19.799999        93.300003
3          18.400000        84.300003
4          17.100000        80.599998

 Define  historical input feature X : the chirps per second and the output feature y: temperature (F) in pandas

In [3]:
X = data[['chirps per second']]
y = data['temperature (F)']
#Note: The double brackets around ‘chirps per second’ are there because we will later be ingesting our X and y into scikit-learn. Scikit-learn can take any number of predictor features X and it expects a 2-D array for X which is why we need the double brackets.

 partition our historical data into a train set and a test set

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42,
test_size = 0.25)

Check the size of your whole data and the size of training and test sets using the shape function.

In [5]:
print('Whole Data shape', data.shape)
print('X_train shape', X_train.shape)
print('X_test shape', X_test.shape)

Whole Data shape (15, 2)
X_train shape (11, 1)
X_test shape (4, 1)


Single Linear Regression (SLR) model is to instantiate it.
This creates a blank model object (algorithm) which is much like an empty template ready to be filled (fitted)
with our training data.

In [6]:
from sklearn.linear_model import LinearRegression
slr_model = LinearRegression()

## Fit the Model on Train Set:

In [7]:
slr_model.fit(X_train, y_train)

LinearRegression()

The slope and the y-intercept for  model and then view a scatterplot with the actual regression line

In [8]:
#To see the simple linear regression parameters
slope = slr_model.coef_
y_intercept = slr_model.intercept_
print('Slope', slope)
print('Intercept', y_intercept)

Slope [3.5571068]
Intercept 20.405003967581408


## Make Predictions on Train Set

In [9]:
y_pred_train = slr_model.predict(X_train)

Compare our actual y from the train set (y_train) to the predicted values
of y ( y_pred_train) by making use of the R2 metric and the MSE metrics

In [15]:
from sklearn.metrics import r2_score, mean_squared_error
r2_train = r2_score(y_train, y_pred_train)
mse_train = mean_squared_error(y_train, y_pred_train)
print('Training R2 score', r2_train)
print('Training MSE', mse_train)

Training R2 score 0.6962423950559918
Training MSE 13.765094164156034


The training **R2** value is 0.70. This means that the model we just built explains about 70% of the variability in our
data. The training **MSE** is 13.77 which means our predictions on average deviate from the true values of y by about
3.71 degrees of Fahrenheit. The MSE is the mean *squared* error, so to get the actual error with the same units as
our **y** (deg. Fahrenheit) we need to take the ***square root of the MSE***.

## Make predictions on the test data

In [12]:
y_pred_test = slr_model.predict(X_test)

### Evaluate the model’s performance on Test Set

In [13]:
r2_test = r2_score(y_test, y_pred_test)
mse_test = mean_squared_error(y_test, y_pred_test)
print('Test R2 score', r2_test)
print('Test MSE', mse_test)

Test R2 score -0.9409079579988209
Test MSE 11.092282161560394


Compare the test instances y_test and y_pred_test in comparison_df.

In [16]:
Comparison_df = pd.DataFrame({'Actual' : y_test, 'Predicted' : y_pred_test})
Comparison_df.to_csv(r'/content/Comparison_df.csv', index=True)
Comparison_df

Actual  Predicted
9   83.300003  78.030137
11  82.599998  81.587244
0   88.599998  91.547140
13  83.500000  80.875820

# Classification Predictive Models

There are multiple types of Logistic regression algorithms.
<ol><li>Binary Logistic Regression: The target variable has only
two possible outcomes, such as Spam or Not Spam, Cancer or No Cancer.</li>
<li>Multinomial Logistic Regression: The target variable has three or more nominal categories, such as predicting
the type of Wine.</li>
<li>Ordinal Logistic Regression: the target variable has three or more ordinal categories, such as restaurant or
product rating from 1 to 5.</li></ol>

In [17]:
#import pandas
import pandas as pd
# load dataset
pima = pd.read_csv('/content/diabetes.csv')
pima.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

Two types of variables dependent(or target
variable) and independent variable(or feature variables).

In [18]:
#split dataset in features and target variable
feature_cols = ['Pregnancies', 'Insulin',
'Age','Glucose','BMI',
'DiabetesPedigreeFunction']
X = pima[feature_cols] # Features
y = pima.Outcome # Target variable